src/main/java/org/opentripplanner/gtfs/mapping/TransferMapper.java

```diff
@@ -173,64 +156,90 @@ private TransferConstraint mapConstraint(Transfer rhs, Trip fromTrip, Trip toTri
+          org.onebusaway.gtfs.model.Stop rhsStopOrStation,
+          org.onebusaway.gtfs.model.Route rhsRoute,
+          Trip trip,
+          boolean boardTrip
   ) {
-    Collection<TransferPoint> result = new ArrayList<>();
-    if (trip != null) {
-      result.addAll(createTransferPointForTrip(stops, trip, TripTransferPoint::new));
-    }
-    else if (route != null) {
-      /*
-      TODO - This code result in a OutOfMemory exception, fin out why and fix it
-           - See issue https://github.com/opentripplanner/OpenTripPlanner/issues/3429
-      for (Trip tripInRoute : tripsByRoute.get(route)) {
-        result.addAll(
-            createTransferPointForTrip(
-              stops,
-              tripInRoute,
-              (t,i) -> new RouteTransferPoint(route, t, i)
-            )
-        );
-      }
-       */
+    Route route = routeMapper.map(rhsRoute);
+    Station station = null;
+    Stop stop = null;
+
+    // A transfer is specified using Stops and/or Station, according to the GTFS specification:
+    //
+    //    If the stop ID refers to a station that contains multiple stops, this transfer rule
+    //    applies to all stops in that station.
+    //
+    // Source: https://developers.google.com/transit/gtfs/reference/transfers-file
+
+    if (rhsStopOrStation.getLocationType() == 0) {
+      stop  = stopMapper.map(rhsStopOrStation);
     }
     else {
-      for (var stop : stops) {
-        result.add(new StopTransferPoint(stop));
-      }
+      station = stationMapper.map(rhsStopOrStation);
+    }
+    if(trip != null) {
+      int stopPositionInPattern = stopPosition(trip, stop, station, boardTrip);
+      return stopPositionInPattern < 0 ? null : new TripTransferPoint(trip, stopPositionInPattern);
+    }
+    else if(route != null) {
+      var trips = tripsByRoute.get(route);
+      if(trips.isEmpty()) { throw new IllegalStateException("No trips found fr route: " + route); }
 
+      int stopPositionInPattern = stopPosition(route, stop, station, boardTrip);
+      return new RouteTransferPoint(route, stopPositionInPattern);
+    }
+    else if(stop != null) {
+      return new StopTransferPoint(stop);
+    }
+    else if(station != null) {
+      return new StationTransferPoint(station);
     }
-    return result;
+
+    throw new IllegalStateException("Should not get here!");
   }
 
-  private Collection<TransferPoint> createTransferPointForTrip(
-      Collection<StopLocation> stops,
-      Trip trip,
-      BiFunction<Trip, Integer, TransferPoint> createPoint
-  ) {
-    Collection<TransferPoint> result = new ArrayList<>();
+
+  private int stopPosition(Route route, Stop stop, Station station, boolean boardTrip) {
+    var stopPosList = tripsByRoute.get(route).stream()
+            .map(t -> stopPosition(t, stop, station, boardTrip))
+            .distinct()
+            .collect(Collectors.toList());
+
+    if(stopPosList.size() == 1) { return stopPosList.get(0); }
+
+    FIXED_ROUTE_ERROR.error(
```

Comment:
Should this log an error or add an issue to the issue store?

Reply:
My mistake, this code is not in use. I have deleted it.